---
title: "Assume teosinte is ancestral"
output: NULL
author: Jinliang Yang
date: 07-04-2020
---


```{r setup, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE) 
knitr::opts_knit$set(root.dir=normalizePath('../../')) 
#library(tidyverse)
```


Note, require `R/3.6` or above for the following step:

Assume teosinte is ancestral
- define 1 as derived state (which can be meth or unmeth)
- run for each genomic feature

```{r}

get_feature_dsfs <- function(pops_file="largedata/mr_100bp/pop3_CpG_matrix_imp.csv",
                   features=c("gene", "exon", "five", "three"), cols=4:54, 
                   ftbed="data/genomic_feature.bed", outpath="cache/gf_sfs_interpop_"){
    ## ftbed: feature file in bed format: headers [seqname, start, end, attribute]
    ## cols to calculate: 
    
    df <- fread(pops_file, data.table=FALSE)
    
    ########### methylation file
    chr <- df[, 1:3]
    chr$start <- as.numeric(as.character(df$pos)) - 100+1
    chr$end <- as.numeric(as.character(df$pos))
    gr1 = with(chr, GRanges(seqnames=chr, IRanges(start=start, end=end), strand="+", uid=uid))
    
    ######### feature file in bed
    bed <- read.table(ftbed)
   
    names(bed) <- c("seqname", "start", "end", "attribute")
    bed$start <- bed$start + 1
     
    for(fi in features){
      myf <- subset(bed, attribute %in% fi)
      gr0 = with(myf, GRanges(seqnames=seqname, IRanges(start=start, end=end), strand="+", geneid=attribute ))
      feature <- gr0
      ex1 = findOverlaps(feature, gr1)
      out <- gr1[subjectHits(ex1)]
    
      mydf <- subset(df, uid %in% mcols(out)$uid)
      message(sprintf("###>>> [%s] rows and [%s] cols for feature [ %s ], sfs...", nrow(mydf), ncol(mydf), fi))
      #########
      
      # ancestral allele status
      mydf$s0 <- apply(mydf[, 4:23], 1, mean)
    
      # avg ancestral < 0.3: unmethylated
      mydf1 <- subset(mydf, s0 < 0.3)
      ftb1 <- apply(mydf1[, cols], 1, getcount1)
      
      # avg ancestral > 0.7: methylated
      mydf2 <- subset(mydf, s0 > 0.7)
      ftb2 <- apply(mydf2[, cols], 1, getcount2)
      
      ftb <- c(ft1, ftb2)
      sfs <- table(ftb)
      outfile <- paste0(outpath, fi, ".csv")
      write.table(sfs, outfile, sep=",", row.names=FALSE, quote=FALSE)
    }
    
}
## ancestral is 0: unmethylated
getcount1 <- function(x, mmin=0.3, mmax=0.7){
    #n0 <- sum(x < mmin) #0
    n2 <- sum(x > mmax) #2 => methylated is derived allele
    n1 <- sum(x >= mmin & x <= mmax) #1
    return(2*n2+n1)
}

## ancestral is 1: methylated
getcount2 <- function(x, mmin=0.3, mmax=0.7){
    n0 <- sum(x < mmin) #0 => unmethylated is the derived alleles
    #n2 <- sum(x > mmax) #2
    n1 <- sum(x >= mmin & x <= mmax) #1
    return(2*n0+n1) # number of derived alleles
}



library("GenomicRanges")
library("data.table")

# ob <- load("largedata/AGPv4_genomic_features.RData")

# CG
get_feature_dsfs(pops_file="largedata/mr_100bp/pop3_CpG_matrix_imp.csv",
                   features=c("Exon","Gene_down_5k","Gene_up_5k","Intergenic","Intron","TE"), cols=4:54, 
                   ftbed="data/genomic_feature_Gen_v4.bed", outpath="largedata/revise_features/dsfs_pop3_gf_CG_")

# CHG  
get_feature_dsfs(pops_file="largedata/mr_100bp/pop3_CHG_matrix_imp.csv",
                features=c("Exon","Gene_down_5k","Gene_up_5k","Intergenic","Intron","TE"),
                ftbed="data/genomic_feature_Gen_v4.bed", outpath="largedata/revise_features/dsfs_pop3_gf_CHG_")
#
```


