---
title: "Genomic feature selection using AGPV4 Annotation"
output: NULL
author: Jinliang Yang
date: 02-07-2019
---


```{r setup, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE) 
knitr::opts_knit$set(root.dir=normalizePath('../../')) 
#library(tidyverse)
```



Note, require `R/3.5` or above for the following step:

```{r}

get_feature_sfs <- function(pops_file="largedata/mr_100bp/pop3_CpG_matrix_imp.csv",
                   features=c("gene", "exon", "five", "three"),
                   f, outpath="cache/gf_sfs_interpop_"){
    
    df <- fread(pops_file, data.table=FALSE)
    
    ###########
    chr <- df[, 1:3]
    chr$start <- as.numeric(as.character(df$pos)) - 100+1
    chr$end <- as.numeric(as.character(df$pos))
    gr1 = with(chr, GRanges(seqnames=chr, IRanges(start=start, end=end), strand="+", uid=uid))
    
    
    for(fi in features){
      gr0 = with(f[[fi]], GRanges(seqnames=seqname, IRanges(start=start, end=end), strand="+", geneid=attribute ))
      feature <- gr0
      ex1 = findOverlaps(feature, gr1)
      out <- gr1[subjectHits(ex1)]
    
      mydf <- subset(df, uid %in% mcols(out)$uid)
      message(sprintf("###>>> [%s] rows and [%s] cols for feature [%s], sfs...", nrow(mydf), ncol(mydf), fi))
      #########
      #df <- as.data.frame(df)
      ftb <- apply(mydf[, -1:-3], 1, getcount)
      sfs <- table(ftb)
      outfile <- paste0(outpath, fi, ".csv")
      write.table(sfs, outfile, sep=",", row.names=FALSE, quote=FALSE)
    }
    
}

getcount <- function(x, mmin=0.3, mmax=0.7){
    n0 <- sum(x < mmin) #0
    n2 <- sum(x > mmax) #2
    n1 <- sum(x >= mmin & x <= mmax) #1
    return(2*n2+n1)
}


library("GenomicRanges")
library("data.table")

ob <- load("largedata/AGPv4_genomic_features.RData")

# CG
get_feature_sfs(pops_file="largedata/mr_100bp/pop3_CpG_matrix_imp.csv",
                   features=c("gene", "exon", "five", "three"),
                   f, outpath="cache/features/gf_sfs_interpop_CG_")

# CHG  
get_feature_sfs(pops_file="largedata/mr_100bp/pop3_CHG_matrix_imp.csv",
                   features=c("gene", "exon", "five", "three"),
                   f, outpath="cache/features/gf_sfs_interpop_CHG_")

# CHH  
get_feature_sfs(pops_file="largedata/mr_100bp/pop3_CHH_matrix_imp.csv",
                   features=c("gene", "exon", "five", "three"),
                   f, outpath="cache/features/gf_sfs_interpop_CHH_")
  

```


### fix 0 sites


### CHH need to be modified:
```{r}
files <- list.files(path="cache/features", pattern = "gf_sfs_interpop_CHH", full.names = T)
files <- files[grep(".csv", files)]

cg <- read.csv("cache/features/gf_sfs_interpop_CG_exon.csv")
for(i in 1:length(files)){
  afile <- read.csv(files[i])
  chh <- merge(cg, afile, by="ftb", all.x=T)
  
  chh[is.na(chh[, 3]), 3] <- 0
  chh <- chh[, c(1,3)]
  names(chh) <- c("ftb", "Freq")
  write.table(chh, files[i], sep=",", row.names=FALSE, quote=FALSE)
}

```


------------------------------
hypo and hyper: Gene, up5k and down5k

```{r}
library("GenomicRanges")
library("data.table")

ob <- load("largedata/AGPv4_gf_hyper_hypo_updown5k.RData")
# CG
get_feature_sfs(pops_file="largedata/mr_100bp/pop3_CpG_matrix_imp.csv",
                   features=c("hyper_gene", "hyper_up", "hyper_down", "hypo_gene","hypo_up","hypo_down"),
                   f, outpath="cache/features/gf3_sfs_interpop_CG_")

# CHG  
get_feature_sfs(pops_file="largedata/mr_100bp/pop3_CHG_matrix_imp.csv",
                   features=c("hyper_gene", "hyper_up", "hyper_down", "hypo_gene","hypo_up","hypo_down"),
                   f, outpath="cache/features/gf3_sfs_interpop_CHG_")

# CHH 
get_feature_sfs(pops_file="largedata/mr_100bp/pop3_CHH_matrix_imp.csv",
                   features=c("hyper_gene", "hyper_up", "hyper_down", "hypo_gene","hypo_up","hypo_down"),
                   f, outpath="cache/features/gf3_sfs_interpop_CHH_")
```

# fix 0 sites


### CHH need to be modified:
```{r}
files <- list.files(path="cache/features/", pattern = "gf3_sfs_interpop_CHH", full.names = T)
files <- files[grep(".csv", files)]
cg <- read.csv("cache/features/gf3_sfs_interpop_CG_hypo_down.csv")
for(i in 1:length(files)){
  afile <- read.csv(files[i])
  chh <- merge(cg, afile, by="ftb", all.x=T)
  
  chh[is.na(chh[, 3]), 3] <- 0
  chh <- chh[, c(1,3)]
  names(chh) <- c("ftb", "Freq")
  write.table(chh, files[i], sep=",", row.names=FALSE, quote=FALSE)
}

```








------------------------------
deprecated!
Gene, up5k and down5k

```{r}
library("GenomicRanges")
library("data.table")

ob <- load("largedata/AGPv4_genomic_features_updown5k.RData")
# CG
get_feature_sfs(pop1="largedata/mr_100bp/teosinte_CpG_matrix_imp.csv", 
                   pop2="largedata/mr_100bp/landrace_CpG_matrix_imp.csv",
                   pop3="largedata/mr_100bp/elite_rm_CpG_matrix_imp.csv",
                   features=c("gene", "up", "down"),
                   f, outpath="largedata/msfs/gf2_sfs_interpop_CG_")

# CHG  
get_feature_sfs(pop1="largedata/mr_100bp/teosinte_CHG_matrix_imp.csv", 
                   pop2="largedata/mr_100bp/landrace_CHG_matrix_imp.csv",
                   pop3="largedata/mr_100bp/elite_rm_CHG_matrix_imp.csv",
                   features=c("gene", "up", "down"),
                   f, outpath="largedata/msfs/gf2_sfs_interpop_CHG_")

# CHH 
get_feature_sfs(pop1="largedata/mr_100bp/teosinte_CHH_matrix_imp.csv", 
                   pop2="largedata/mr_100bp/landrace_CHH_matrix_imp.csv",
                   pop3="largedata/mr_100bp/elite_rm_CHH_matrix_imp.csv",
                   features=c("gene", "up", "down"),
                   f, outpath="largedata/msfs/gf2_sfs_interpop_CHH_")
```

# fix 0 sites


### CHH need to be modified:
```{r}
files <- list.files(path="largedata/msfs", pattern = "gf2_sfs_interpop_CHH", full.names = T)

cg <- read.csv("largedata/msfs/gf2_sfs_interpop_CG_down.csv")
for(i in 1:length(files)){
  afile <- read.csv(files[i])
  chh <- merge(cg, afile, by="ftb", all.x=T)
  
  chh[is.na(chh[, 3]), 3] <- 0
  chh <- chh[, c(1,3)]
  names(chh) <- c("ftb", "Freq")
  write.table(chh, files[i], sep=",", row.names=FALSE, quote=FALSE)
}

```