---
title: "Plot par figures"
output: NULL
author: Jinliang Yang
date: 02-26-2019
---


```{r setup, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE) 
knitr::opts_knit$set(root.dir=normalizePath('../../')) 
#library(tidyverse)
```



### Teosinte Genome-wide results

```{r}
### Jinliang Yang
### 10-14-2016
### run MCMC for comet on chr10

source("lib/mplots.R")
source("lib/mcmcbc.R")

get_numus <- function(obfile="cache/mcmc_res/new_msfs_CG_pops.RData", burnin=0.2, Ne=150000){
    
    ob <- load(obfile)
    
    s.samples <- res[['samples']]$s
    nu.samples <- res[['samples']]$nu
    mu.samples <- res[['samples']]$mu
    l.samples <- res[['samples']]$l
    mu.acc <- res[['acc']]$mu
    nu.acc <- res[['acc']]$nu
    s.acc <- res[['acc']]$s
    
    s.samples=s.samples[(length(s.samples)*burnin+1):length(s.samples)]
    nu.samples=nu.samples[(length(nu.samples)*burnin+1):length(nu.samples)]
    mu.samples=mu.samples[(length(mu.samples)*burnin+1):length(mu.samples)]
    
    tem1 <- data.frame(mn=mean(Ne*s.samples), sd=sd(Ne*s.samples), med=median(Ne*s.samples), type="s")
    tem2 <- data.frame(mn=mean(-log10(nu.samples)), sd=sd(-log10(nu.samples)), med=median(-log10(nu.samples)), type="nu")
    tem3 <- data.frame(mn=mean(-log10(mu.samples)), sd=sd(-log10(mu.samples)), med=median(-log10(mu.samples)), type="mu")
    
    out <- rbind(tem1, tem2, tem3)
    return(out)
}

### files
files <- list.files("largedata/revise_features", pattern="RData", full.names = T)

out <- data.frame()
for(i in 1:length(files)){
  ne <- gsub("_ft.*", "", files[i])
  ne <- as.numeric(as.character(gsub(".*_", "", ne)))
  df11 <- get_numus(obfile=files[i], burnin=0.2, Ne=ne)
  df11$par <- files[i]
  df11$ne <- ne
  out <- rbind(out, df11)
}

out$class <- gsub(".*\\/|_pop3.*", "", out$par)


out$context <- gsub(".*pop3_|.*gf_", "", out$par)
out$ft <- gsub(".*G_", "", out$context)
out$context <- gsub("_.*", "", out$context)
out$ft <- gsub("_[0-9].*_ft.RData", "", out$ft)

write.table(out, "cache/response_ne_feature.csv", sep=",", row.names=FALSE, quote=FALSE)

```

## Feature seperately

```{r}
### files
files <- list.files("largedata/revise_ft_sep", pattern="RData", full.names = T)

out <- data.frame()
for(i in 1:length(files)){
  ne <- gsub("_ft.RData", "", files[i])
  ne <- as.numeric(as.character(gsub(".*_", "", ne)))
  df11 <- get_numus(obfile=files[i], burnin=0.2, Ne=ne)
  df11$par <- files[i]
  df11$ne <- ne
  out <- rbind(out, df11)
}

out$class <- gsub(".*\\/|_pop3.*", "", out$par)
out$teo <- gsub(".*teo-", "", out$par)
out$teo <- gsub("_.*", "", out$teo)

out$context <- gsub(".*pop3_|.*gf_", "", out$par)
out$ft <- gsub(".*G_", "", out$context)
out$context <- gsub("_.*", "", out$context)
out$ft <- gsub("_teo.*", "", out$ft)

write.table(out, "cache/response_sep_feature.csv", sep=",", row.names=FALSE, quote=FALSE)

```



### plot the results:


```{r}
library("tidyr")
library("ggplot2")
df <- read.csv("cache/response_sep_feature.csv")

#df <- gather(data=df0, key=k, value="p", 1:3)


fsize=18
df$ne <- factor(df$ne, levels = c("50000", "1e+05", "5e+05", "1e+06"))
df$ft <- factor(df$ft, levels = c("Exon","Intron", "Gene_up_5k","Gene_down_5k", "Intergenic", "TE" ), 
                labels=c("Exon","Intron", "Up5k", "Down5k", "Intergenic", "TE" ))

df1 <- subset(df, teo %in% "hyper")


p1 <- ggplot(df1, aes(x=type, y=mn, fill=ne)) +
    geom_bar(stat="identity", position=position_dodge()) +
    facet_grid( ft ~ context) +
    xlab("Posterior Estimates") +
    ylab("Value") +
    geom_errorbar(aes(ymin=mn-sd, ymax=mn+sd), width=.2,
                 position=position_dodge(.9)) +
    # https://www.color-hex.com/color-palette/73961
    scale_fill_manual(values=c("#cc4141", "#72114b", "#e1cd07", "#5f9057"),
                      name=c("Ne"),
                      labels=c("5e+4", "1e+5", "5e+5", "1e+6")) +
    #scale_fill_manual(values=c("#cc4141", "#72114b", "#e1cd07", "#5f9057", "#114c63")) +
    #labs(fill = "Dose (mg)") +
    scale_x_discrete(labels=c("-log10(mu)","-log10(nu)", "Ne*s")) +
    theme(legend.position = "top", plot.title = element_text(size=20), axis.text=element_text(size=15),
          strip.text.x = element_text(size = 20, face = "bold"),
          strip.text.y = element_text(size = 15, face = "bold"),
          axis.title=element_text(size=fsize, face="bold"),
                legend.title = element_text(size=fsize, face="bold"),
                legend.text = element_text(size=fsize))
  #########

p1

```

```{r}
df2 <- subset(df, teo %in% "hypo")

p2 <- ggplot(df2, aes(x=type, y=mn, fill=ne)) +
    geom_bar(stat="identity", position=position_dodge()) +
    facet_grid( ft ~ context) +
    xlab("Posterior Estimates") +
    ylab("Value") +
    geom_errorbar(aes(ymin=mn-sd, ymax=mn+sd), width=.2,
                 position=position_dodge(.9)) +
    # https://www.color-hex.com/color-palette/73961
    scale_fill_manual(values=c("#cc4141", "#72114b", "#e1cd07", "#5f9057"),
                      name=c("Ne"),
                      labels=c("5e+4", "1e+5", "5e+5", "1e+6")) +
    #scale_fill_manual(values=c("#cc4141", "#72114b", "#e1cd07", "#5f9057", "#114c63")) +
    #labs(fill = "Dose (mg)") +
    scale_x_discrete(labels=c("-log10(mu)","-log10(nu)", "Ne*s")) +
    theme(legend.position = "top", plot.title = element_text(size=20), axis.text=element_text(size=15),
          strip.text.x = element_text(size = 20, face = "bold"),
          strip.text.y = element_text(size = 15, face = "bold"),
          axis.title=element_text(size=fsize, face="bold"),
                legend.title = element_text(size=fsize, face="bold"),
                legend.text = element_text(size=fsize))
  #########

p2

```



```{r}
library("cowplot")
pdf("graphs/SFig_Ne_feature_teo-hyper.pdf", width=8, height=8)
plot_grid(p1, ncol=1, labels = NULL, label_size = 26)
dev.off()

pdf("graphs/SFig_Ne_feature_teo-hypo.pdf", width=8, height=8)
plot_grid(p2, ncol=1, labels = NULL, label_size = 26)
dev.off()
```


```{r}
tem <- subset(df[, -2], type == "s")
df0 <- ddply(tem, .(context, teo, ft), summarise,
             d = sd(mn, na.rm=T),
             m = mean(mn))

df0$teo <- factor(df0$teo, levels = c("hyper", "hypo"), labels =c("Hyper", "Hypo"))
p <- ggplot(df0, aes(x=ft, y=m, fill=teo)) +
    geom_bar(stat="identity", position=position_dodge()) +
    facet_grid( ~ context) +
    xlab("") +
    ylab("Selection (Ne x s)") +
    geom_errorbar(aes(ymin=m-d, ymax=m+d), width=.2,
                 position=position_dodge(.9)) +
    # https://www.color-hex.com/color-palette/73961
    scale_fill_manual(values=c("#E69F00", "#56B4E9"),
                      name=c("Ancestral status")) +
                      #labels=c("Teo Hypo", "Teo Hyper")) +
    #scale_fill_manual(values=c("#cc4141", "#72114b", "#e1cd07", "#5f9057", "#114c63")) +
    #labs(fill = "Dose (mg)") +
    #scale_x_discrete(labels=c("-log10(mu)","-log10(nu)", "Ne*s")) +
    theme_classic() +
    theme(legend.position = c(0.8,0.8), plot.title = element_text(size=20), 
          axis.text=element_text(size=12, face="bold"),
          strip.text.x = element_text(size = 20, face = "bold"),
          #strip.text.y = element_text(size = 15, face = "bold"),
          axis.title=element_text(size=fsize, face="bold"),
                legend.title = element_text(size=fsize, face="bold"),
                legend.text = element_text(size=fsize)) 
p
```
```{r}
library("cowplot")
pdf("graphs/SFig_feature_teo_nes.pdf", width=10, height=4)
plot_grid(p, ncol=1, labels = NULL, label_size = 26)
dev.off()

```

```{r}
tb2 <- ddply(df, .(context, teo, type, ft), summarise,
            val = mean(mn))
tb2$rate <- 10^(-tb2$val)

t <- tb2[order(tb2$val),]
head(subset(t, type == "mu" & context == "CG"))

head(subset(t, type == "nu" & context == "CG"))
```

