---
title: "Run VCAP through huskeR for Ranodm sets"
output: html_notebook
---

## VCAP using BED file

The BED file genome coordinates should be 0-based with the chromStart inclusive and the chromEnd exclusive (see https://genome.ucsc.edu/FAQ/FAQformat.html#format1). Note that Ensembl genome database, gff3, gtf, and Tassel5 coordinates are 1-based (and gff3 and gtf ranges are start and end-inclusive). A simple way to go from Ensemble/gff3/gtf/Tassel coordinates to BED file coordinates without subtracting 1 is to think of the BED file coordinates as 1-based but with start exclusive and end inclusive. Hence, a BED file range with chromStart = 5000 and chromEnd = 5100 includes chromosomal positions ≥ 5001 and ≤ 5100 in terms of a Tassel genotype file. The coordinate system (i.e., AGPv3) should be consistent between the BED and genotype files.


```{r, eval=FALSE}

####################
get_asb_feature_random <- function(myasb, mybg, ntimes=10,
                                   outputbase="largedata/VCAP/bg_K1_rep" ){
  
  for(i in 1:ntimes){
    idx <- sample(1:nrow(mybg), nrow(myasb))
  
    bg1 <- mybg[idx, c("CHR", "pos", "pos")]
    names(bg1) <- c("chr", "start", "end")
    bg1$start <- bg1$start - 1
    message(sprintf("###>>> feature sampling BG SNPs [%s]", nrow(bg1)))
    write.table(bg1, paste0(outputbase, i, ".bed"), sep="\t", 
            row.names=FALSE, col.names=FALSE, quote=FALSE)
  }
}

frq <- read.csv("cache/ASB_BG_frq.csv")
asb <- subset(frq, type %in% "ASB")
bg <- subset(frq, type %in% "bg")
  

# K1) promotor_5kb;
myasb <- subset(asb, promoter_5kb %in% "yes")
mybg <- subset(bg, promoter_5kb %in% "yes")
get_asb_feature_random(myasb, mybg, ntimes=10, outputbase="largedata/random_kin/bg_K1_rep" )

# K2) promotor_1kb + 5' UTR;
p1k5 <- subset(asb, promoter_1kb %in% "yes" | five_prime_UTR %in% "yes")
mybg <- subset(bg, promoter_1kb %in% "yes" | five_prime_UTR %in% "yes")
get_asb_feature_random(myasb=p1k5, mybg, ntimes=10, outputbase="largedata/random_kin/bg_K2_rep" )

# K3) Exon + Intron;
exin <- subset(asb, exon %in% "yes" | intron %in% "yes")
mybg <- subset(bg, exon %in% "yes" | intron %in% "yes")
get_asb_feature_random(myasb=exin, mybg, ntimes=10, outputbase="largedata/random_kin/bg_K3_rep" )

# K4) 3' UTR + post_gene_1kb
post1k <- subset(asb, post_gene_1kb %in% "yes" | three_prime_UTR %in% "yes")
mybg <- subset(bg, post_gene_1kb %in% "yes" | three_prime_UTR %in% "yes")
get_asb_feature_random(myasb=post1k, mybg, ntimes=10, outputbase="largedata/random_kin/bg_K4_rep" )

# K5) far from genes (non_genic)
nong <- subset(asb, non_genic %in% "yes")
mybg <- subset(bg, non_genic %in% "yes")
get_asb_feature_random(myasb=nong, mybg, ntimes=10, outputbase="largedata/random_kin/bg_K5_rep" )

```

### Run VCAP via huskeR

```{r, eval=FALSE}

library("huskeR")
df <- data.frame(bedfile1=list.files(path="largedata/random_kin", pattern="bg_K1_rep.*bed$", full.names=TRUE),
                 bedfile2=list.files(path="largedata/random_kin", pattern="bg_K2_rep.*bed$", full.names=TRUE),
                 bedfile3=list.files(path="largedata/random_kin", pattern="bg_K3_rep.*bed$", full.names=TRUE),
                 bedfile4=list.files(path="largedata/random_kin", pattern="bg_K4_rep.*bed$", full.names=TRUE),
                 bedfile5=list.files(path="largedata/random_kin", pattern="bg_K5_rep.*bed$", full.names=TRUE),
                 gz.lix="/lustre/work/jyanglab/jyang21/dbcenter/RareAlleles/genomeAnnos/VCAP/genotypes/NAM/namrils_projected_hmp31_MAF02mnCnt2500.hmp.txt.gz.lix",
                 genome_kinship="/lustre/work/jyanglab/jyang21/dbcenter/RareAlleles/genomeAnnos/VCAP/kinship/NAM_HM31_MAF02mnCnt2500_WholeGenomeCentered_IBS_SqrMatrix.txt",
                 res_kinship="largedata/random_kin/asb_res_kinship"
                 )

df$res_kinship <- df$bedfile5
df$res_kinship <- gsub("K5", "res", df$res_kinship)
df$res_kinship <- gsub(".bed", "", df$res_kinship)

run_VCAP(df, email = "yangjl0930@gmail.com", jobid = "run_vcap", runinfo = c(TRUE,
       "jclarke", "10", "50G", "3:00:00"))

```

### Run LDAK

Using R package `huskeR`:

```{r}
library("huskeR")

### get kinship list
kfile <- list.files(path="largedata/random_kin", pattern="N.bin", full.names=TRUE)

for(i in 1:10){
  idx <- grep(paste0("rep", i, "\\."), kfile)
  d <- data.frame(kin=gsub(".grm.N.bin", "", kfile[idx]))
  write.table(d, paste0("largedata/random_kin/bg_", i, "_klist.txt"), sep="\t", 
            row.names=FALSE, col.names=FALSE, quote=FALSE)
}


rdf <- data.frame()
###
for(i in 1:10){
  pheno <- list.files(path="/lustre/work/jyanglab/jyang21/dbcenter/RareAlleles/genomeAnnos/VCAP/phenotypes/NAM/familyCorrected", pattern="NAM", full.names=TRUE)
  df <- data.frame(output="largedata/h2_bg/bg_", klist=paste0("largedata/random_kin/bg_", i, "_klist.txt"), pheno=pheno)
  df$output <- paste0(df$output, i, "_", gsub(".*\\/|.txt", "", df$pheno))
  rdf <- rbind(rdf, df)
}

set_ldak(rdf, email="yangjl0930@gmail.com", runinfo=c(TRUE, "jclarke", "2", "10G", "8:00:00"))

```


----------------------------



## get Results

```{r}
files <- list.files(path="largedata/h2_bg", pattern="reml", full.names=TRUE)

out <- data.frame()
for(i in 1:length(files)){
  h2 <- read.table(files[i], skip=13, header=TRUE)
  th2 <- as.data.frame(t(h2[-7, 1:2]))
  names(th2) <- as.character(h2$Component[-7])
  th2$trait <- files[i]
  out <- rbind(out, th2[-1, ])
}

out$trait <- gsub(".*bg_|_famC.*", "", out$trait)

#out$Her_ALL <- as.numeric(as.character(out$Her_ALL))
#out1 <- subset(out, Her_ALL > 0.4)
write.table(out, "cache/h2_bg_updated.csv", sep=",", row.names=FALSE, quote=FALSE)
```


