---
title: "Run VCAP through huskeR"
output: html_notebook
date: 02-22-2019
author: "Jinliang Yang"
---


### Maize and Teosinte CG

02192022 version:
five mutually exclusive features:
1. gene body
2. downstream 5k
3. upstream 5k
4. intergenic
5. TE

```{r, eval=FALSE}


library("huskeR")
run_batch_vcap_sample <- function(sid=1:10,
  input_dir = "/common/jyanglab/shared/Gen_Xu/DMR_Features_Overlap_02192022",
  dmr_patterns = c("CG_teosinte_maize", "CHG_teosinte_maize", "CHH_teosinte_maize",)){
  
  # setup file system and find the dmr pattern
  files <- list.files(input_dir, full.names = TRUE, pattern="bed")
  df <- data.frame(f=files, pop="1")
  
  ### loop over patterns
  outdf <- data.frame()
  for(i in 1:length(dmr_patterns)){
    idx <- grep(dmr_patterns[i], df$f)
    temdf <- df[idx, ]
    temdf$fld <- gsub(".bed", "_sample", temdf$f)
    temdf$f <- gsub(".*\\/|.bed", "", temdf$f)
    temdf$f <- paste0(temdf$fld, "/", temdf$f, "_S")
    
    #outdf <- data.frame(f=rep(temdf$f, each=length(sid)), smpl = rep(sid, times=nrow(temdf)))
    #outdf$f <- paste0(outdf$f, outdf$smpl, ".bed")
    
    outdf1 <- data.frame()
    for(j in 1:length(sid)){
      # copy the files to the new folder
      dir.create(paste0("largedata/VCAP/", dmr_patterns[i], "/sample"), showWarnings = FALSE)
      
      output_dir <- paste0("largedata/VCAP/", dmr_patterns[i], "/sample/S", sid[j])
      dir.create(output_dir, showWarnings = FALSE)
      dir.create(paste0(output_dir, "/bedfile"), showWarnings = FALSE)
      
      temdf$out <- paste0(temdf$f, sid[j], ".bed")
      
      file.copy(as.character(temdf$out), paste0(output_dir, "/bedfile"))
      temdf$files <- gsub(".*\\/", "", temdf$out)
      temdf$files <- paste0(output_dir, "/bedfile/", temdf$files)
    
      temp <- data.frame(bedfile1=temdf$files[1],
                 bedfile2=temdf$files[2],
                 bedfile3=temdf$files[3],
                 bedfile4=temdf$files[4],
                 bedfile5=temdf$files[5],
                 res = paste0(output_dir, "/bedfile/rest_geno"))
      outdf1 <- rbind(outdf1, temp)
    }
    
    outdf <- rbind(outdf, outdf1)
  }
  
  
  ### df for batch job input:
  df <- data.frame(bedfile1=outdf$bedfile1,
                 bedfile2=outdf$bedfile2,
                 bedfile3=outdf$bedfile3,
                 bedfile4=outdf$bedfile4,
                 bedfile5=outdf$bedfile5,
                 gz.lix="/common/jyanglab/shared/dbcenter/RareAlleles/genomeAnnos/VCAP/genotypes/NAM/namrils_projected_hmp31_MAF02mnCnt2500.hmp.txt.gz.lix",
                 genome_kinship="/common/jyanglab/shared/dbcenter/RareAlleles/genomeAnnos/VCAP/kinship/NAM_HM31_MAF02mnCnt2500_WholeGenomeCentered_IBS_SqrMatrix.txt",
                 res_kinship=outdf$res
                 )

  run_VCAP(df, email = "yangjl0930@gmail.com", jobid = "run_vcap", runinfo = c(TRUE,"jclarke", "5", "20G", "3:00:00"))
}

run_batch_vcap_sample(sid=1:100,
  input_dir = "/common/jyanglab/shared/Gen_Xu/DMR_Features_Overlap_02192022",
  dmr_patterns = "CG_teosinte_maize")


run_batch_vcap_sample(sid=1:100,
  input_dir = "/common/jyanglab/shared/Gen_Xu/DMR_Features_Overlap_02192022",
  dmr_patterns = c("CG_teosinte_maize", "CHG_teosinte_maize", "CHH_teosinte_maize",
                  "CG_landrace_maize", "CHG_landrace_maize", "CHH_landrace_maize",
                  "CG_teosinte_landrace", "CHG_teosinte_landrace", "CHH_teosinte_landrace"))

```

### Run LDAK

Using R package `huskeR`:

```{r, eval=FALSE}
library("huskeR", lib.loc = "~/R/x86_64-pc-linux-gnu-library/3.5")

run_batch_ldak_sample <- function(sid=1:10, traits,
  dmr_patterns = c("CG_teosinte_maize", "CHG_teosinte_maize", "CHH_teosinte_maize",
                  "CG_landrace_maize", "CHG_landrace_maize", "CHH_landrace_maize",
                  "CG_teosinte_landrace", "CHG_teosinte_landrace", "CHH_teosinte_landrace")
  ){
  
  df <- data.frame()
  ###
  pheno <- list.files(path="/common/jyanglab/shared/dbcenter/RareAlleles/genomeAnnos/VCAP/phenotypes/NAM/familyCorrected", pattern="NAM", full.names=TRUE)
  ph <- data.frame(pheno, trait=1)
  ph$trait <- gsub(".*/NAM_|_famCorrected.txt", "", ph$pheno)
  ph <- subset(ph, trait %in% traits)
    
  ### get kinship list
  for(i in 1:length(dmr_patterns)){
    
    df1 <- data.frame()
    for(j in 1:length(sid)){
      ### write out kinship list file
      kinship_dir <- paste0("largedata/VCAP/", dmr_patterns[i], "/sample/S", j, "/bedfile")
      
      kfile <- list.files(path=kinship_dir, pattern="N.bin", full.names=TRUE)
      d <- data.frame(kin=gsub(".grm.N.bin", "", kfile))
      write.table(d, paste0(kinship_dir, "/kinlist.txt"), sep="\t", 
            row.names=FALSE, col.names=FALSE, quote=FALSE)
      
      output_h2 <- gsub("bedfile", "h2/", kinship_dir)
      dir.create(output_h2, showWarnings = FALSE)
      tem <- data.frame(output=output_h2, klist=paste0(kinship_dir, "/kinlist.txt"), pheno=pheno)
      tem$output <- paste0(tem$output, gsub(".*\\/|.txt", "", tem$pheno))
      df1 <- rbind(df1, tem)
    }
    df <- rbind(df, df1)
  }
  
  run_ldak(df, email="yangjl0930@gmail.com", runinfo=c(TRUE, "jyanglab", "2", "10G", "8:00:00"))
}

res <- read.csv("cache/h2_teo-maize_CG.csv")
traits <- as.character(res$trait)[1]
run_batch_ldak_sample(sid=1:100, dmr_patterns = "CG_teosinte_maize")

```


