---
title: "VCFv4 teo20 from Wenbin"
output: html_notebook
---

# download VCF AGPV4 from Wenbin:

```{bash}
#rsync -alPvz farm:/group/jrigrp4/VCF_JRIAL2_20samples/ .
irsync -rKV VCF_JRIAL2_20samples/ i:/iplant/home/yangjl/
irsync -rKV i:/iplant/home/yangjl/BS_teo/teo20_VCF_AGPv4 /lustre/work/jyanglab/jyang21/hugedata/teo20_VCF_AGPv4/
# QD < 2.0 || FS > 60.0 || MQ < 40.0 || MQRankSum < -12.5 || ReadPosRankSum < -8.0

# get fastq data to iplant
nohup irsync -rKV i:/iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-7_R1.fastq.gz /common/jyanglab/shared/dbcenter/teosinte_20parents/ & 
nohup irsync -rKV i:/iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-7_R2.fastq.gz /common/jyanglab/shared/dbcenter/teosinte_20parents/ &

nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-7_R1.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-7_R2.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-8_R1.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-8_R2.fastq.gz & 

nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-1_R1.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-1_R2.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-2_R1.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-2_R2.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-3_R1.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-3_R2.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-4_R1.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-4_R2.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-5_R1.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-5_R2.fastq.gz & 

nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-6_R1.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-6_R2.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-9_R1.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-9_R2.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-10_R1.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-10_R2.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-11_R1.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-11_R2.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-12_R1.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-12_R2.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-13_R1.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-13_R2.fastq.gz & 


nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-14_R1.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-14_R2.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-15_R1.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-15_R2.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-16_R1.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-16_R2.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-17_R1.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-17_R2.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-18_R1.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-18_R2.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-19_R1.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-19_R2.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-20_R1.fastq.gz & 
nohup iget -P /iplant/home/rossibarra/rare_alleles_seq/teosinte_20parents/JRIAL2/JRIAL2_fastq/JRIAL2-20_R2.fastq.gz & 




# get data back to Gen BAM
nohup irsync -rKV i:/iplant/home/yangjl/BS_teo/bam_dedup /common/jyanglab/shared/dbcenter/BS_teo/bam_dedup/ & 

# get data back to Gen fastq
irsync -rKV i:/iplant/home/yangjl/BS_teo/WGBS_fastq_lz4 /common/jyanglab/shared/dbcenter/BS_teo/WGBS_fastq

```


## my `bash_alias` file

```{bash, eval=FALSE}
# -----------------------
# HPC system alias
# ----------------------
# our own computational resource 
# -p jyanglab

alias ll='ls -lF -h'
alias mysq='squeue --format="%.18i %.9A %.9P %.8j %.8u %.2t %.10M %.6D %R %.9m"'
alias srun1='srun --nodes=1 --mem 1024 --ntasks=1 --licenses=common --time=8:00:00 --pty bash'
alias srun4='srun --nodes=1 --mem 4G --ntasks=4 --licenses=common --time=8:00:00 --pty bash'
alias srunq4='srun --qos=short --nodes=1 --licenses=common --ntasks=4 --mem 32G --time 6:00:00 --pty bash'
alias srunq1="srun --qos=short --nodes=1 --licenses=common --ntasks-per-node=1 --mem-per-cpu=1024 --time 6:00:00 --pty $SHELL"
alias myjob='mysq | grep jyang21'
alias mylab='mysq | egrep "jyang21|raghu|xie186"'
# Add an "alert" alias for long running commands.  Use like so:
#   sleep 10; alert
alias alert='notify-send --urgency=low -i "$([ $? = 0 ] && echo terminal || echo error)" "$(history|tail -n1|sed -e '\''s/^\s*[0-9]\+\s*//;s/[;&|]\s*alert$//'\'')"'
#
alias squeue='squeue --format="%.12i %.0P %.18j %.8u %.2t %.10M %.6D %R %C"'

```


## Concatenate or combine VCF/BCF files.

Need to index them and then concatenate.

```{bash, eval=FALSE}
## test and then run for chr1-9
module load tabix
tabix -p vcf filtered_JRIAL2_20samples_Chr10_Feb13_2017.vcf.gz
for i in `seq 1 9`; do tabix -p vcf filtered_JRIAL2_20samples_Chr"$i"_Feb13_2017.vcf.gz; done

## run cat
bcftools concat -a --remove-duplicates --file-list input_list.txt --threads 10 --output-type z -o filtered_JRIAL2_20samples_Chrall_Feb13_2017.vcf.gz

## remove indels and keep only variant sites
#bcftools view filtered_JRIAL2_20samples_Chrall_Feb13_2017.vcf.gz --types snps --min-ac 1 --threads 10 --output-type z -o filtered_JRIAL_20samples_Chrall_03202018_snps.vcf.gz.gz

# Use -m2 -M2 -v snps to only view biallelic SNPs. keep only variant sites using --min-ac
bcftools view filtered_JRIAL_20samples_Chrall_Feb13_2017.vcf.gz -m2 -M2 -v snps --min-ac 1 -Oz -o filtered_JRIAL_20samples_Chrall_03202018_bisnps.vcf.gz

# Convert to annotation format (have to be done manually to add the header)
bcftools query -f '%CHROM\t%POS\t%REF\t%ALT[\t%IUPACGT]\n' filtered_JRIAL_20samples_Chrall_03202018_bisnps.vcf.gz > JRI20_bisnps_annot.txt
# bcftools query -f 'chr\tpos\tref\talt[\t%SAMPLE]\n' JRI20_bi_snps_annot.vcf.gz > JRI20_bi_snps_annot.header

```


## Convert to PLINK format

### 1. Set SNP ID for all the data.
```{bash, eval=FALSE}
module load plink
plink -vcf test3_chr10.vcf.gz --keep-allele-order --set-missing-var-ids @_#_\$1_\$2 --make-bed --out chr10
```


### 2. Get quantgen stat

```{bash}
plink -bfile chr10 --freq --out chr10  

plink -bfile chr10 --freq --missing --pca --ibc --out chr10  
# --ibc (ported from GCTA) calculates three inbreeding coefficients for each sample
```
