---
title: "Run Pseduo Ref"
output: html_notebook
---

[PseduoRef](https://github.com/yangjl/pseudoRef)

## SNP tables

```{bash, eval=FALSE}
module load bcftools

# get bi-allelic SNPs
bcftools view filtered_JRIAL_20samples_Chrall_03202018_bisnps.vcf.gz -m2 -M2 -v snps -Oz -o JRI20_bi_snps_annot.vcf.gz

# get SNP table
bcftools query -f '%CHROM\t%POS\t%REF\t%ALT[\t%IUPACGT]\n' filtered_JRIAL_20samples_Chrall_03202018_bisnps.vcf.gz > JRI20_bisnps_annot.txt

## get header
bcftools query -f 'chr\tpos\tref\talt[\t%SAMPLE]\n' filtered_JRIAL_20samples_Chrall_03202018_bisnps.vcf.gz > JRI20_bisnps_annot.header
head -n 1 JRI20_bisnps_annot.header > header

```


##





```{r}
#devtools::install_github("yangjl/pseudoRef")

library("data.table")
library("Biostrings")
library("pseudoRef")




arules <- data.frame(from=c("M", "Y", "R", "K"), to=c("C", "C", "G", "T"))
outdir <- "largedata/pgenome"

h <- read.table("largedata/teo20_VCF_AGPv4/header", header=TRUE)
snpdt <- fread("largedata/teo20_VCF_AGPv4/JRI20_bisnps_annot.txt", header=FALSE, nrows=10)

teo20 <- read.csv("data/teo20_ids.csv")
names(snpdt) <- gsub("_.*", "", names(h))

# fa can be "character", "DNAStringSet" or "DNAString"
fa <- "/project/jyanglab/jyang21/dbcenter/AGP/AGPv4/Zea_mays.AGPv4.dna.toplevel_chr1-10.fa"
fa <- readDNAStringSet(filepath = fa, format="fasta")

pseudoRef(fa, snpdt, sidx = 5:ncol(snpdt), arules, outdir)


# Error in `[.data.frame`(x, i) : object 'SAMPLE' not found

library(farmeR)
inputdf <- data.frame(input.vcf="largedata/gatk_vcf/JRIAL1A_joint_call.filtered_snps.vcf",
                      out.fa="largedata/pgenome/mysample.fa")
run_pseudoref(inputdf,
              ref.fa="$HOME/dbcenter/AGP/AGPv2/Zea_mays.AGPv2.14.dna.toplevel.fa",
              gatkpwd="$HOME/bin/GenomeAnalysisTK-3.5/GenomeAnalysisTK.jar",
              email="yangjl0930@gmail.com", runinfo = c(TRUE, "bigmemh", 8))



### run bsmap
library(farmeR)
inputdf <- data.frame(fq1="/group/jrigrp4/BS_teo20/fastq/JRA1_CTTGTA_R1.fastq.gz",
                      fq2="/group/jrigrp4/BS_teo20/fastq/JRA1_CTTGTA_R2.fastq.gz",
                      out="$HOME/Documents/Github/methylation/largedata/bsmap/JRA1")

runa_bsmap(inputdf, ref.fa="$HOME/Documents/Github/methylation/largedata/pgenome/mysample.fa",
           picardpwd="$HOME/bin/picard-tools-2.1.1/picard.jar",
           email="yangjl0930@gmail.com", runinfo = c(TRUE, "bigmemm", 8))

```



