---
title: "DMR results from Gen"
output: html_notebook
author: Jinliang Yang
date: 09-20-2019
---


```{r setup, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE) 
knitr::opts_knit$set(root.dir=normalizePath('../../')) 
#library(tidyverse)
```

# Modelling

```{r}
## Using R/QTL
d <- read.delim("doc/Table_S_DMRs.csv", header=TRUE, sep=",")
dim(d)
# 2709753

subset(d, chr ==4 & start > 203000000 & stop < 204000000)

gene_dmr <- read.delim("data/all_DMR_overp_genes.txt")
gene_dmr <- gene_dmr[order(abs(gene_dmr$mean_methylation_difference), decreasing = T),]

```

enriched in genic region?

### LD calculation

```{r}
df <- read.table("data/Teo_Lan_Maize_vgt1.txt", header=TRUE)

### Missing rate
df$mr <- apply(df[, -1:-5], 1, function(x) {sum(is.na(x))})
df$raf <- apply(df[, -1:-5], 1, function(x) {
  r1 <- sum(!is.na(x) & x == 1)
  r2 <- sum(!is.na(x) & x == 0)
  r3 <- sum(!is.na(x) & x == -1)
  return((2*r1 + r2)/(2*r1 + 2*r3 + 2*r2))
})

sub <- subset(df, raf < 0.9 & raf > 0.1)
### Reference allele frq

names(sub) <- gsub("_.*", "", names(sub))
names(sub)[3] <- "uid"
```




```{r}
m1 <- read.csv("cache/vgt1_CG.csv")
m2 <- read.csv("cache/vgt1_CG_landrace.csv")
m3 <- read.csv("cache/vgt1_CG_teosinte.csv")

m <- merge(m1[, -c(which(names(m1) %in% c("chr", "pos")))], m2[, -c(which(names(m2) %in% c("chr", "pos")))], by="uid")
m <- merge(m, m3[, -c(which(names(m3) %in% c("chr", "pos")))], by="uid")

names(m) <- gsub(".*_", "", names(m))
#m$type <- "dmr"

#sub$type <- "snp"
sub2 <- sub[, names(m)]

t <- rbind(m, sub2)


out <- sub2[, 1:2]
out$r2 <- apply(sub2[, -1], 1, function(x){
  df <- t(rbind(m[m$uid == "8_135969201_135969300", -1], x)) 
  df <- subset(df, !is.na(df[, 1]) & !is.na(df[,2]))
  return(cor(df[,1], df[,2]) )
  }
  )

out$pos <- as.numeric(as.character(gsub(".*_", "", out$uid)))

```

