---
title: "DMR and HiChIP"
output: html_notebook
author: Jinliang Yang
date: 05-22-2019
---


```{r setup, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE) 
knitr::opts_knit$set(root.dir=normalizePath('../../')) 
#library(tidyverse)
```



## DMR

```{r}
dmr <- read.delim("data/all_DMRs_03052019.txt", header=TRUE)
table(dmr$Compare)

## Teosinte and Maize DMR
tm <- subset(dmr, Compare %in% "Teosinte/Maize")
tm$bp <- abs(tm$stop - tm$start)

hist(log10(tm$bp), breaks = 100, col="#eed5b7", main="DMR ", xlab="log10(bp)")
abline(v=median(log10(tm$bp)), lty=2, lwd=2, col="red")
median(tm$bp) # 315
mean(tm$bp) #436

```


## Play with the DMR master file

```{r}
library("data.table")
library("tidyr")
library("ggplot2")
library("plyr")

d <- fread("data/DMR_gene_feature_overlap_05242019.txt", data.table=FALSE)

tm <- subset(d, DMR_Comparison %in% "Teosinte/Maize")

dtm <- ddply(tm, .(DMR_Context, Feature), summarise,
             size=sum(DMR_Feature_Overlap_size),
             hic=sum(width, na.rm=TRUE))
dtm$nohic <- dtm$size - dtm$hic


#dtm <- gather(tm[, c("DMR_Context", "Feature", "DMR_Feature_Overlap_size")], key="Context", value="size", 3)

ln <- read.table("data/Annotation_file/With_TE_NO_Merge/genomic_feature_length.txt", header=T)

dtm2 <- gather(dtm[, -3], key="HiC", value="size", 3:4)

dtm2 <- merge(dtm2, ln, by="Feature", all.x=TRUE)
dtm2$ns <- dtm2$size/dtm2$bp


p <- ggplot(dtm2, aes(x=DMR_Context, y=ns, fill=HiC)) +
    facet_grid(. ~ Feature) +
    geom_bar(stat="identity", color="black") +
    theme_classic() +
    labs(title="Teo-Maize DMRs", x="", y = "Normalized space (bp 100%)")+
    # https://www.color-hex.com/color-palette/73961
    #scale_fill_manual(values=c("#cc4141", "#72114b", "#e1cd07", "#5f9057", "#114c63", "#d3d3d3"),
    #                  name=c(""),
    #                  labels=c("Gene", "Down 5k", "Up 5k", "Intergenic", "TE", "Rest")) +
    scale_fill_manual(values=c("#ae5a41", "#559e83", '#999999')) +
    #                    name="", labels=c("Teosinte", "Maize")) +
    #scale_x_discrete(labels=c("Hyper","Hypo")) +
    theme(legend.position="top", axis.text=element_text(size=18), axis.title=element_text(size=18) )
  #########
p


```




```{r}
find_overlaps <- function(df1, df2){
    
    ## df1: chr, start, end, ratio (mehtylation ratio) [data.frame]
    ## df2: chr, start, end
    
    #df$chr <- gsub("Chr", "", df$chr)
    
    gr1 <- with(df1, GRanges(seqnames=chr, IRanges(start=start, end=end)))
    gr2 <- with(df2, GRanges(seqnames=chr, IRanges(start=start, end=end)))
    
    #######
    tb <- findOverlaps(query=gr1, subject=gr2)
    tb <- as.matrix(tb)
    
    out1 <- as.data.frame(gr1[tb[,1]])
    out2 <- as.data.frame(gr2[tb[,2]])
    
    myout <- list(out1, out2)
    return(myout)
}

df2 <- hichip
df2 <- data.frame(chr=c(hichip$Chr_1, hichip$Chr_2), 
                  start=c(hichip$S1, hichip$S2),
                  end=c(hichip$E1, hichip$E2))
df2$id <- paste(df2$chr, df2$start, df2$end, sep="_")

## unique anchors
df2 <- df2[!duplicated(df2$id),]
df2$bp <- df2$end - df2$start
sum(df2$bp)

df1 <- tm
names(df1)[1:3] <- c("chr", "start", "end")


out1$uid <- paste(out1$seqnames, out1$start, out1$end, sep="_") 
out1 <- out1[!duplicated(out1$uid),] #536

tm$uid <- paste(tm$DMR_Chr, tm$DMR_Start, tm$DMR_End, sep="_") 
tm <- merge(tm, out1, by="uid", all.x=TRUE)

```





# HiChIP Results


```{r}
hichip <- read.table("data/hichipper_interation.txt", header=TRUE)

sum(hichip$Chr_1 == hichip$Chr_2)
#which(hichip$Chr_1 != hichip$Chr_2)

hichip$len1 <- hichip$E1 - hichip$S1
hichip$len2 <- hichip$E2 - hichip$S2


### The length of the interaction sites
l <- c(hichip$len1, hichip$len2)
hist(log10(l), breaks = 100, col="#eed5b7", main="Length the interaction sites", xlab="Log10(bp)")
abline(v=median(log10(l)), lty=2, lwd=2, col="red")
median(l) 
mean(l)

### The distance of the two interaction sites
hichip$dis <- (hichip$S2 + hichip$E2)/2 - (hichip$S1 + hichip$E1)/2
d <- hichip$dis
hist(log10(d), breaks = 100, col="#eed5b7", main="Distance of the two interaction sites", xlab="Log10(bp)")
abline(v=median(log10(d)), lty=2, lwd=2, col="red")
median(d) 
mean(d)

```



## Final Overalp between DMR and HiChIP

```{r}
library("GenomicRanges")
#library("data.table")
#library(plyr)




find_overlaps <- function(df1, df2){
    
    ## df1: chr, start, end, ratio (mehtylation ratio) [data.frame]
    ## df2: chr, start, end
    
    #df$chr <- gsub("Chr", "", df$chr)
    
    gr1 <- with(df1, GRanges(seqnames=chr, IRanges(start=start, end=end)))
    gr2 <- with(df2, GRanges(seqnames=chr, IRanges(start=start, end=end)))
    
    #######
    tb <- findOverlaps(query=gr1, subject=gr2)
    tb <- as.matrix(tb)
    
    out1 <- as.data.frame(gr1[tb[,1]])
    out2 <- as.data.frame(gr2[tb[,2]])
    
    myout <- list(out1, out2)
    return(myout)
}



#####
names(tm)[4] <- "end"
df1 <- tm


df2 <- hichip
df2 <- data.frame(chr=c(hichip$Chr_1, hichip$Chr_2), 
                  start=c(hichip$S1, hichip$S2),
                  end=c(hichip$E1, hichip$E2))
df2$id <- paste(df2$chr, df2$start, df2$end, sep="_")

## unique anchors
df2 <- df2[!duplicated(df2$id),]
df2$bp <- df2$end - df2$start
sum(df2$bp)

tb <- table(df2$id)

### use the function
out <- find_overlaps(df1, df2)

out1 <- out[[1]]
out1$uid <- paste(out1$seqnames, out1$start, out1$end, sep="_")
df1$uid <- paste(df1$chr, df1$start, df1$end, sep="_")

sub <- subset(df1, uid %in% out1$uid)


### plot
library("tidyr")
library("ggplot2")

d <- gather(sub[, c("Type", "POPA", "POPB")], key="pop", value="mr", 2:3)

p1 <- ggplot2(d, )

p <- ggplot(d, aes(x=Type, y=mr, fill=pop)) +
    #facet_grid(. ~ type) +
    geom_violin(trim=FALSE) +
    geom_boxplot(width=0.1, position=position_dodge(0.9))+
    theme_classic() +
    labs(title="HiChIP anchors in the DMR", x="Context", y = "Methylation level")+
    # https://www.color-hex.com/color-palette/73961
    #scale_fill_manual(values=c("#cc4141", "#72114b", "#e1cd07", "#5f9057", "#114c63", "#d3d3d3"),
    #                  name=c(""),
    #                  labels=c("Gene", "Down 5k", "Up 5k", "Intergenic", "TE", "Rest")) +
    scale_fill_manual(values=c("#ae5a41", "#559e83"),
                        name="", labels=c("Teosinte", "Maize")) +
    #scale_x_discrete(labels=c("Hyper","Hypo")) +
    theme(legend.position="top", axis.text=element_text(size=18), axis.title=element_text(size=18) )
  #########
p
```


